### PR TITLE
feat: add [gitlab] contributors service

### PR DESCRIPTION
**Pull request by sunny0826**
re #8077 

/assign @calebcartwright

### CODE REVIEW

File: services/gitlab/gitlab-contributors.service.js
Comment (chris48s) [inline services/gitlab/gitlab-contributors.service.js:70]: Even though we are extracting the value from the header instead of the response body we should still validate the value exists and makes sense before we pass the value to `renderContributorBadge()`. You can either do it by applying a Joi schema, or do some checks inline here and manually throw an `InvalidResponse` if needed https://github.com/badges/shields/blob/c73072deed8eb6a5b9f51a06377f5f396ef21842/core/base-service/errors.js#L80

**Comment by chris48s** [comment]
Thanks for working on this. A few general points:

* For some reason, Circle CI isn't running on this PR or https://github.com/badges/shields/pull/8083 but it looks like it worked on https://github.com/badges/shields/pull/8076 . Any idea what changed? I'd like to see if we can ge this resolved particularly if we have multiple PRs to review from you. Feel free to open another PR if we can find a fix. We can transfer the review over if needed

* When we submit a PR that adds or modifies service code, we put the service name in square brackets in the PR title so we can run the relevant service tests in the CI build. See docs at https://github.com/badges/shields/blob/7d452472defa0e0bd71d6443393e522e8457f856/CONTRIBUTING.md#running-service-tests-in-pull-requests . That's why Caleb changed the title of https://github.com/badges/shields/pull/8076 I've edited this one too, although at the moment it is moot as CI is not running on PRs from your fork.

* Can you avoid tagging individual maintainers for review on PRs that could be reviewed by anyone on the core team. I don't think it absolutely has to be Caleb that reviews this specific PR so lets not generate notification noise for people if we don't need to.

File: services/gitlab/gitlab-contributors.service.js
Comment (sunny0826) [reply services/gitlab/gitlab-contributors.service.js]: done

**Comment by sunny0826** [comment]
> * For some reason, Circle CI isn't running on this PR or [fix: gitlab licence service docs and example #8083](https://github.com/badges/shields/pull/8083) but it looks like it worked on [fix(gitlab service test): fix gitlab pipeline & tag(nested group) service test, run [gitlabtag gitlabpipeline] #8076](https://github.com/badges/shields/pull/8076) . Any idea what changed? I'd like to see if we can ge this resolved particularly if we have multiple PRs to review from you. Feel free to open another PR if we can find a fix. We can transfer the review over if needed

This issue has been with me since the [first PR](https://github.com/badges/shields/pull/8024) I submitted. Unfortunately, I have not used Circle CI, but this documentation may be helpful.
https://circleci.com/docs/2.0/workflows/#workflows-waiting-for-status-in-github

> * When we submit a PR that adds or modifies service code, we put the service name in square brackets in the PR title so we can run the relevant service tests in the CI build. See docs at https://github.com/badges/shields/blob/7d452472defa0e0bd71d6443393e522e8457f856/CONTRIBUTING.md#running-service-tests-in-pull-requests . That's why Caleb changed the title of [fix(gitlab service test): fix gitlab pipeline & tag(nested group) service test, run [gitlabtag gitlabpipeline] #8076](https://github.com/badges/shields/pull/8076) I've edited this one too, although at the moment it is moot as CI is not running on PRs from your fork.

Got it. I will pay attention to this issue in subsequent PRs.

> * Can you avoid tagging individual maintainers for review on PRs that could be reviewed by anyone on the core team. I don't think it absolutely has to be Caleb that reviews this specific PR so lets not generate notification noise for people if we don't need to.

I'm sorry about that. I just think Caleb understands the context of this PR better. I'm still new to shields and he has helped me a lot before.

File: services/gitlab/gitlab-contributors.service.js
Comment (chris48s) [inline services/gitlab/gitlab-contributors.service.js:76]: I think you've got the wrong idea here.

If we're not using any values from the body, there is no reason to validate the contents of the body.

The purpose of validating is to make sure the value we need exists and makes sense so if we get a response that doesn't have an `x-total` header we aren't trying to call `metric(undefined)` or something like that.

So we could write something like

```js
import { nonNegativeInteger } from '../validators.js'

const schema = Joi.object({ 'x-total': nonNegativeInteger }).required()

async handle(...
  // ...

  const data = this.constructor._validate(res.headers, schema)
  const contributorCount = data['x-total']
  return this.constructor.render({ contributorCount })
}
```

here to

- make sure the headers contain the value we want
- it is in the expected format

There are some docs on how and why we validate at https://github.com/badges/shields/blob/master/doc/input-validation.md hopefully that helps.

File: services/gitlab/gitlab-contributors.service.js
Comment (chris48s) [inline services/gitlab/gitlab-contributors.service.js:81]: Can you clarify what you're going for with `contributorCount = +parsed.last.page` here?

**Comment by chris48s** [comment]
> I'm sorry about that. I just think Caleb understands the context of this PR better. I'm still new to shields and he has helped me a lot before.

No worries - its not a huge deal but I did want to flag it for future :)

File: services/gitlab/gitlab-contributors.service.js
Comment (sunny0826) [reply services/gitlab/gitlab-contributors.service.js:76]: Oh, I did misunderstand, I thought I needed to validate the response content. Thanks for your guidance, I will fix it.

File: services/gitlab/gitlab-contributors.service.js
Comment (sunny0826) [reply services/gitlab/gitlab-contributors.service.js:81]: Use the number of pages instead of the number of contributors because setting the `options: { searchParams: { page: '1', per_page: '1' } },`

File: services/gitlab/gitlab-contributors.tester.js
Comment (chris48s) [inline services/gitlab/gitlab-contributors.tester.js:16]: Service code is looking good now :+1: 

In terms of the tests, we don't ideally want to test individual values at the service level. This one will just break if a second person contributes to this repo and it is not really testing anything beyond what the test above using `isMetric`.

With the "5k" test: `metric()` itself has its own test suite https://github.com/badges/shields/blob/8f7164f96bcdb097c310797974541b45d33ed58c/services/text-formatters.spec.js#L40 so we don't really need to re-test the functionality of this in every badge that uses it.

I think we can drop down to just the

- 'Contributors'
- 'Contributors (repo not found)'
- 'Mocking the missing x-total header'

tests and ditch the other 2.

Aside from that, I think this is good to merge. Thanks

Comment (sunny0826) [reply services/gitlab/gitlab-contributors.tester.js:16]: To prevent misunderstanding, you mean to delete the following two test cases：
- `1 contributor`
-  `Mocking 5k contributors`

Is it right?

**Review by chris48s — APPROVED** [review]

**Comment by chris48s** [comment]
thanks for your work on this

**Pull request merged**
